File: src-front/adapters/shared/market-data.ts
// Market Data Commands
import type {
  SymbolSearchResult,
  Asset,
  Quote,
  UpdateAssetProfile,
  MarketDataProviderInfo,
} from "@/lib/types";
import type { QuoteImport } from "@/lib/types/quote-import";
import type { MarketDataProviderSetting } from "../types";

import { invoke, logger } from "./platform";

export const searchTicker = async (query: string): Promise<SymbolSearchResult[]> => {
  try {
    return await invoke<SymbolSearchResult[]>("search_symbol", { query });
  } catch (error) {
    logger.error("Error searching for ticker.");
    throw error;
  }
};

export const syncHistoryQuotes = async (): Promise<void> => {
  try {
    await invoke<void>("synch_quotes");
  } catch (error) {
    logger.error("Error syncing history quotes.");
    throw error;
  }
};

export const getAssetProfile = async (assetId: string): Promise<Asset> => {
  try {
    return await invoke<Asset>("get_asset_profile", { assetId });
  } catch (error) {
    logger.error("Error loading asset data.");
    throw error;
  }
};

export const getAssets = async (): Promise<Asset[]> => {
  try {
    return await invoke<Asset[]>("get_assets");
  } catch (error) {
    logger.error("Error loading assets.");
    throw error;
  }
};

export const getLatestQuotes = async (assetIds: string[]): Promise<Record<string, Quote>> => {
  try {
    return await invoke<Record<string, Quote>>("get_latest_quotes", { assetIds });
  } catch (error) {
    logger.error("Error loading latest quotes.");
    throw error;
  }
};

export const updateAssetProfile = async (payload: UpdateAssetProfile): Promise<Asset> => {
  try {
    // Internal transformation: add id from payload.symbol
    return await invoke<Asset>("update_asset_profile", { id: payload.symbol, payload });
  } catch (error) {
    logger.error("Error updating asset profile.");
    throw error;
  }
};

export const deleteAsset = async (id: string): Promise<void> => {
  try {
    await invoke<void>("delete_asset", { id });
  } catch (error) {
    logger.error("Error deleting asset.");
    throw error;
  }
};

export const updatePricingMode = async (assetId: string, pricingMode: string): Promise<Asset> => {
  try {
    return await invoke<Asset>("update_pricing_mode", { id: assetId, pricingMode });
  } catch (error) {
    logger.error("Error updating asset pricing mode.");
    throw error;
  }
};

export const updateQuote = async (symbol: string, quote: Quote): Promise<void> => {
  try {
    return await invoke<void>("update_quote", { symbol, quote });
  } catch (error) {
    logger.error("Error updating quote");
    throw error;
  }
};

export const syncMarketData = async (
  assetIds: string[],
  refetchAll: boolean,
  refetchRecentDays?: number,
): Promise<void> => {
  try {
    await invoke<void>("sync_market_data", { assetIds, refetchAll, refetchRecentDays });
  } catch (error) {
    logger.error(`Error refreshing quotes for assets: ${String(error)}`);
    throw error;
  }
};

export const deleteQuote = async (id: string): Promise<void> => {
  try {
    return await invoke<void>("delete_quote", { id });
  } catch (error) {
    logger.error("Error deleting quote");
    throw error;
  }
};

export const getQuoteHistory = async (symbol: string): Promise<Quote[]> => {
  try {
    return await invoke<Quote[]>("get_quote_history", { symbol });
  } catch (error) {
    logger.error(`Error fetching quote history for symbol ${symbol}.`);
    throw error;
  }
};

export const getMarketDataProviders = async (): Promise<MarketDataProviderInfo[]> => {
  try {
    return await invoke<MarketDataProviderInfo[]>("get_market_data_providers");
  } catch (error) {
    logger.error("Error fetching market data providers.");
    throw error;
  }
};

export const getMarketDataProviderSettings = async (): Promise<MarketDataProviderSetting[]> => {
  try {
    return await invoke<MarketDataProviderSetting[]>("get_market_data_providers_settings");
  } catch (error) {
    logger.error("Error fetching market data provider settings.");
    throw error;
  }
};

export const updateMarketDataProviderSettings = async (payload: {
  providerId: string;
  priority: number;
  enabled: boolean;
}): Promise<MarketDataProviderSetting> => {
  try {
    return await invoke<MarketDataProviderSetting>("update_market_data_provider_settings", payload);
  } catch (error) {
    logger.error("Error updating market data provider settings.");
    throw error;
  }
};

export const checkQuotesImport = async (
  file: File,
  hasHeaderRow = true,
): Promise<QuoteImport[]> => {
  try {
    const buffer = await file.arrayBuffer();
    const content = Array.from(new Uint8Array(buffer));
    return await invoke<QuoteImport[]>("check_quotes_import", { content, hasHeaderRow });
  } catch (error) {
    logger.error("Error checking quotes import.");
    throw error;
  }
};

export const importManualQuotes = async (quotes: QuoteImport[]): Promise<QuoteImport[]> => {
  try {
    // Internal transformation: hardcode overwriteExisting flag
    const overwriteExisting = true;
    return await invoke<QuoteImport[]>("import_quotes_csv", { quotes, overwriteExisting });
  } catch (error) {
    logger.error("Error importing manual quotes.");
    throw error;
  }
};

File: apps/tauri/src/commands/market_data.rs
use std::collections::HashMap;
use std::sync::Arc;

use crate::{
    context::ServiceContext,
    events::{emit_portfolio_trigger_recalculate, emit_portfolio_trigger_update, PortfolioRequestPayload},
};

use log::{debug, error};
use tauri::{AppHandle, State};
use wealthfolio_core::quotes::{
    service::ProviderInfo, MarketSyncMode, Quote, QuoteImport, SymbolSearchResult,
};
use wealthfolio_market_data::ExchangeInfo;

#[tauri::command]
pub async fn search_symbol(
    query: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<SymbolSearchResult>, String> {
    state
        .quote_service()
        .search_symbol(&query)
        .await
        .map_err(|e| format!("Failed to search ticker: {}", e))
}

#[tauri::command]
pub async fn sync_market_data(
    asset_ids: Option<Vec<String>>,
    refetch_all: bool,
    refetch_recent_days: Option<i64>,
    handle: AppHandle,
) -> Result<(), String> {
    // Determine the appropriate market sync mode based on refetch_all flag
    let market_sync_mode = if let Some(days) = refetch_recent_days {
        MarketSyncMode::RefetchRecent { asset_ids, days }
    } else if refetch_all {
        MarketSyncMode::BackfillHistory {
            asset_ids,
            days: 365 * 5, // 5 years of history as fallback
        }
    } else {
        MarketSyncMode::Incremental { asset_ids }
    };

    let payload = PortfolioRequestPayload::builder()
        .account_ids(None)
        .market_sync_mode(market_sync_mode)
        .build();
    emit_portfolio_trigger_update(&handle, payload);
    Ok(())
}

#[tauri::command]
pub async fn update_quote(
    quote: Quote,
    state: State<'_, Arc<ServiceContext>>,
    handle: AppHandle,
) -> Result<(), String> {
    debug!("Updating quote: {:?}", quote);
    state
        .quote_service()
        .update_quote(quote.clone())
        .await
        .map(|_| ())
        .map_err(|e| e.to_string())?;

    // Manual quote update - no market sync needed, but force full recalculation
    // so historical valuations are recomputed with the updated quotes
    let handle = handle.clone();
    tauri::async_runtime::spawn(async move {
        let payload = PortfolioRequestPayload::builder()
            .account_ids(None)
            .market_sync_mode(MarketSyncMode::None)
            .build();
        emit_portfolio_trigger_recalculate(&handle, payload);
    });
    Ok(())
}

#[tauri::command]
pub async fn delete_quote(
    id: String,
    state: State<'_, Arc<ServiceContext>>,
    handle: AppHandle,
) -> Result<(), String> {
    debug!("Deleting quote: {}", id);
    state
        .quote_service()
        .delete_quote(&id)
        .await
        .map_err(|e| e.to_string())?;

    // Manual quote deletion - no market sync needed, but force full recalculation
    // so historical valuations are recomputed without the deleted quotes
    let handle = handle.clone();
    tauri::async_runtime::spawn(async move {
        let payload = PortfolioRequestPayload::builder()
            .account_ids(None)
            .market_sync_mode(MarketSyncMode::None)
            .build();
        emit_portfolio_trigger_recalculate(&handle, payload);
    });
    Ok(())
}

#[tauri::command]
pub async fn get_quote_history(
    symbol: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<Quote>, String> {
    debug!("Fetching quote history for symbol: {}", symbol);
    state
        .quote_service()
        .get_historical_quotes(&symbol)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_latest_quotes(
    asset_ids: Vec<String>,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<HashMap<String, Quote>, String> {
    state
        .quote_service()
        .get_latest_quotes(&asset_ids)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_market_data_providers(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<ProviderInfo>, String> {
    debug!("Received request to get market data providers");
    state
        .quote_service()
        .get_providers_info()
        .await
        .map_err(|e| {
            error!("Failed to get market data providers: {}", e);
            e.to_string()
        })
}

#[tauri::command]
pub async fn check_quotes_import(
    content: Vec<u8>,
    has_header_row: bool,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<QuoteImport>, String> {
    debug!(
        "Checking quotes import from {} bytes CSV (has_header={})",
        content.len(),
        has_header_row
    );
    state
        .quote_service()
        .check_quotes_import(&content, has_header_row)
        .await
        .map_err(|e| {
            error!("Failed to check quotes import: {}", e);
            format!("Failed to check quotes import: {}", e)
        })
}

#[tauri::command]
pub async fn import_quotes_csv(
    quotes: Vec<QuoteImport>,
    overwrite_existing: bool,
    state: State<'_, Arc<ServiceContext>>,
    handle: AppHandle,
) -> Result<Vec<QuoteImport>, String> {
    debug!(
        "Importing {} quotes from CSV (overwrite_existing={})",
        quotes.len(),
        overwrite_existing
    );
    let result = state
        .quote_service()
        .import_quotes(quotes, overwrite_existing)
        .await
        .map_err(|e| {
            error!("TAURI COMMAND: import_quotes_csv failed: {}", e);
            format!("Failed to import CSV quotes: {}", e)
        })?;

    // Quote import - no market sync needed, just recalculate
    let handle = handle.clone();
    tauri::async_runtime::spawn(async move {
        debug!("Triggering portfolio recalculation after quote import");
        let payload = PortfolioRequestPayload::builder()
            .account_ids(None)
            .market_sync_mode(MarketSyncMode::None)
            .build();
        emit_portfolio_trigger_recalculate(&handle, payload);
    });

    Ok(result)
}

#[tauri::command]
pub fn get_exchanges() -> Vec<ExchangeInfo> {
    wealthfolio_market_data::get_exchange_list()
}
